### PR TITLE
8256172: Clean up CDS handling of i2i_entry

### PR DESCRIPTION
**Pull request by iklam**
This PR is a prerequisite for [JDK-8250989](https://bugs.openjdk.java.net/browse/JDK-8250989) - Consolidate buffer allocation code for CDS static/dynamic archive dumping.

-  Move the allocation of the i2i buffer after all classes are loaded. This makes it possible to estimate the size of the CDS archive before we allocate the output space (this is what dynamic archive does now).
- No need to generate the i2i trampoline code during -Xshare:dump
- Clean up the CDS code in abstractInterpreter.cpp and add more comments.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256172](https://bugs.openjdk.java.net/browse/JDK-8256172): Clean up CDS handling of i2i_entry

### Reviewers
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1280/head:pull/1280`
`$ git checkout pull/1280`

### CODE REVIEW

File: src/hotspot/share/memory/metaspaceShared.cpp [hunk 1]
```diff
@@ -510,19 +507,19 @@ void MetaspaceShared::serialize(SerializeClosure* soc) {
...
-    }
-  } else if (UseSharedSpaces) {
-    assert(_i2i_entry_code_buffers != NULL, "must already been initialized");
-  } else {
-    return NULL;
-  }
+void MetaspaceShared::init_misc_code_space() {
+  // We don't want any valid object to be at the very bottom of the archive.
+  // See ArchivePtrMarker::mark_pointer().
+  MetaspaceShared::misc_code_space_alloc(16);
```
Comment (yminqi) [inline src/hotspot/share/memory/metaspaceShared.cpp:513]: 16 looks arbitrary number though we know it is minimal object alignment --- maybe we we can use object_alignment instead? up to you.

**Review by yminqi — APPROVED** [review]
Look good to me.

**Review by calvinccheung — APPROVED** [review]
Cleanup looks good.

**Comment by iklam** [comment]
Thanks @calvinccheung and @yminqi for the review.
/integrate

**Pull request closed without merging**